src/backend/catalog/test/storage_pending_deletes_redo_test.c

```diff
@@ -0,0 +1,1120 @@
+
+	PdlRedoRemoveTree(xid_to_remove, NULL, 0);
+
+	PdlRedoDropFiles();
+
+	assert_int_equal(DropRelationFiles_call_count, TEST_EXPECTED_NOTES_COUNT - 1);
+
+	/*
+	 * Check that data for removed xids is not touched by PdlRedoDropFiles and
+	 * replace it with InvalidOid to simplify further check.
+	 */
+	int			idx = xid_to_remove - ShmemVariableCache->oldestXid;
+
+	assert_int_equal(test_expected_relnodes[idx].relNode, TEST_REL_OID1 + idx);
+	test_expected_relnodes[idx].relNode = InvalidOid;
+
+	/*
+	 * Check that all other are replaced with InvalidOid.
+	 */
+	for (int i = 0; i < TEST_EXPECTED_NOTES_COUNT; i++)
+	{
+		assert_int_equal(test_expected_relnodes[i].relNode, InvalidOid);
+	}
+}
+
+/*
+ * Scenario:
+ * add several pending delete node with the different xids
+ * and remove pending deletes for one of the xids + some sub_xids
+ * and then drop files.
+ */
+static void
+test_12(void **state)
+{
+	setup(12);
+	ShmemVariableCache->oldestXid = (TransactionId) 1;
+
+	for (int i = 0; i < TEST_EXPECTED_NOTES_COUNT; i++)
+	{
+		PendingRelXactDelete pd =
+		{
+			/* add oldest xid here just to ensure that all nodes will be added */
+			.xid = ShmemVariableCache->oldestXid + (TransactionId) i,
+			.relnode.isTempRelation = false,
+			.relnode.node.spcNode = TEST_TABLESPACE_OID1,
+			.relnode.node.dbNode = TEST_DB_OID1,
+			.relnode.node.relNode = TEST_REL_OID1 + i
+		};
+
+		PdlRedoAdd(&pd);
+
+		/* and fill data which is expected... */
+		test_expected_relnodes[i] = pd.relnode.node;
+	}
+
+	TransactionId xid_to_remove = 5;
+	TransactionId sub_xids_to_remove[] = {10, 11, 12, 15};
+	int			nsubxacts = sizeof(sub_xids_to_remove) / sizeof(TransactionId);
+
+	PdlRedoRemoveTree(xid_to_remove, sub_xids_to_remove, nsubxacts);
+
+	PdlRedoDropFiles();
+
+	assert_int_equal(DropRelationFiles_call_count, TEST_EXPECTED_NOTES_COUNT - 5);
+
+	/*
+	 * Check that data for removed xids is not touched by PdlRedoDropFiles and
+	 * replace it with InvalidOid to simplify further check...
+	 */
+	int			idx = xid_to_remove - ShmemVariableCache->oldestXid;
+
+	assert_int_equal(test_expected_relnodes[idx].relNode, TEST_REL_OID1 + idx);
+	test_expected_relnodes[idx].relNode = InvalidOid;
+	/* ...including all subtransactions. */
+	for (int j = 0; j < nsubxacts; j++)
+	{
+		idx = sub_xids_to_remove[j] - ShmemVariableCache->oldestXid;
+		assert_int_equal(test_expected_relnodes[idx].relNode, TEST_REL_OID1 + idx);
+		test_expected_relnodes[idx].relNode = InvalidOid;
+	}
+
+	/*
+	 * Check that now all expected nodes are replaced with InvalidOid.
+	 */
+	for (int i = 0; i < TEST_EXPECTED_NOTES_COUNT; i++)
+	{
+		assert_int_equal(test_expected_relnodes[i].relNode, InvalidOid);
+	}
+}
+
+static XLogRecord *
+test_create_xlog_record(int pending_deletes_count)
+{
+	int			buffer_size =
+	SizeOfXLogRecord + sizeof(Size) + sizeof(PendingRelXactDelete) * pending_deletes_count;
+
+	char	   *buffer = palloc0(buffer_size);
+	XLogRecord *record = (XLogRecord *) buffer;
+
+	return record;
```

Comment:
the buffer variable is redundant
```suggestion
	return (XLogRecord *) palloc0(buffer_size);
```